[docs] Adding the guide on database initialization from tests (#2273) (#2286)
vividus-framework/vividus: docs/modules/guides/pages/use-in-memory-db-to-manage-test-data.adoc

implementation(group: 'com.h2database', name: 'h2', version: '2.0.202')
----

. Configure connection to the H2 database and initialization
+

> You could execute SQL on connection via the INIT property or through the step to execute SQL query.

Reviewer: ```suggestion
DDL/DQL scripts could be run using either `INIT` property of the connection string or the step executing SQL queries.
```

> Whichever method of initializing a table is chosen, you could use RUNSCRIPT commands with the path to SQL files. Here you can specify absolute path to file or use the 'classpath:/' prefix - files will be searched from 'src/main/resources' folder relatively.

Reviewer: ```suggestion
Whichever method of tables initialization is chosen, `RUNSCRIPT` statement with the path to SQL files could be used. Here you can specify absolute path to file or use the 'classpath:/' prefix - files will be searched from 'src/main/resources' folder relatively.
```

[NOTE]
The details on executing SQL statements on connection
https://h2database.com/html/features.html#execute_sql_on_connection[could be found in the official H2 guide].
+

> .. Using INIT property

Reviewer: ```suggestion
.. Using `INIT` property
```

>  - using INIT property data loaded into the database once and can no longer be change.

Reviewer: ```suggestion
 - using `INIT` property data loaded into the database once and can no longer be change.
```

+
.Configure connection and database initialization
[source,properties]
----
db.connection.testh2db.driver-class-name=org.h2.Driver
db.connection.testh2db.url=jdbc:h2:mem:test;\
  INIT=create schema if not exists test\\;\
  SET SCHEMA test\\;\
  DROP TABLE IF EXISTS test.students\\;\
  CREATE TABLE test.students (firstname VARCHAR(50) NOT NULL, lastname VARCHAR(50) NOT NULL)\\;\
  INSERT INTO students (firstname, lastname) VALUES ('Maria', 'Miller')\\;
db.connection.testh2db.username=
db.connection.testh2db.password=
----
+
.Configure connection and database initialization with RUNSCRIPT
[source,properties]
----
db.connection.testh2db.driver-class-name=org.h2.Driver
db.connection.testh2db.url=jdbc:h2:mem:test;\
  INIT=create schema if not exists test\\;\
  SET SCHEMA test\\;\
  RUNSCRIPT FROM 'classpath:/path/to/sql/file1.sql'\\;\
  RUNSCRIPT FROM 'classpath:/path/to/sql/file2.sql'
db.connection.testh2db.username=
db.connection.testh2db.password=
----
.. Using test step

>  - if you use table initialization in the step, you only need to specify connection parameters in the properties file and initialize the table in the test.

Reviewer: ```suggestion
 - if you use tables initialization in the step, you only need to specify the connection parameters in the properties file and execute DDL script in the test.
```

+
.Configure connection to database
[source,properties]
----
db.connection.testh2db1.driver-class-name=org.h2.Driver
db.connection.testh2db1.url=jdbc:h2:mem:test;\
  DB_CLOSE_DELAY=-1;
db.connection.testh2db1.username=
db.connection.testh2db1.password=
----

> WARNING: If you are using table initialization in the step, do not forget to set the property http://www.h2database.com/html/features.html#in_memory_databases[DB_CLOSE_DELAY].

Reviewer: ```suggestion
WARNING: If you are using table initialization in the step, do not forget to set the property http://www.h2database.com/html/features.html#in_memory_databases[`DB_CLOSE_DELAY`].
```

+

> .Initialization database

Reviewer: ```suggestion
.Initialization of the database
```

[source,gherkin]
----
When I execute SQL query `
  CREATE SCHEMA IF NOT EXISTS test;
  SET SCHEMA test;
  CREATE TABLE test.students (firstname VARCHAR(50) NOT NULL, lastname VARCHAR(50) NOT NULL);
  INSERT INTO students (firstname, lastname) VALUES ('Maria', 'Miller');
` against `testh2db1`
----
+

> .Initialization database with RUNSCRIPT

Reviewer: ```suggestion
.Initialization of database with `RUNSCRIPT`
```

[source,gherkin]
----
When I execute SQL query `
  CREATE SCHEMA IF NOT EXISTS test;
  SET SCHEMA test;
  RUNSCRIPT FROM 'classpath:/path/to/sql/file1.sql';
  RUNSCRIPT FROM 'classpath:/path/to/sql/file2.sql'
` against `testh2db1`

> ----
> . Use the configured connection and the initialized database to execute SQL

Reviewer: ```suggestion
----

. Use the configured connection and the initialized database to execute SQL
```

SELECT * FROM students
` against `testh2db` and save result to scenario variable `h2-records`
----
+

> .Execute SQL against H2 database if the table was created in the step with the schema 'test'

Reviewer: ```suggestion
.Execute DML script against H2 database initialized in the step with the schema `test`
```

[source,gherkin]
----
When I execute SQL query `

> SELECT * FROM test.students

Reviewer: it makes sense to put `INSERT`, `UPDATE` or any other DML statement
